kong/core/balancer.lua

```diff
@@ -1,41 +1,224 @@
+local function load_targets_into_memory(upstream_id)
+  ngx_log(ngx_DEBUG, "fetching targets for upstream: ",tostring(upstream_id))
+  local target_history, err = singletons.dao.targets:find_all {upstream_id = upstream_id}
+  if err then
+    return nil, err
   end
-  target.ip = ip
-  target.port = port
-  return true
+  
+  -- some raw data updates
+  for _, target in ipairs(target_history) do
+    -- split `target` field into `name` and `port`
+    local port
+    target.name, port = string.match(target.target, "^(.-):(%d+)$")
+    target.port = tonumber(port)
+    -- need exact order, so order by created time and uuid
+    target.order = target.created_at..":"..target.id
+  end
+  
+  -- order by time
+  table.sort(target_history, function(a,b) return a.order<b.order end)
+--for i, t in ipairs(target_history) do
+--  ngx_log(ngx_DEBUG, i, t.order)
+--end
+  return target_history
 end
 
-local retry_dns = function(target)
-  local ip, port = toip(target.upstream.host, target.upstream.port, true)
-  if type(ip) ~= "string" then
-    return nil, port
+-- applies the history of lb transactions from index `start` forward
+-- @param rb ring-balancer object
+-- @param history list of targets/transactions to be applied
+-- @param start the index where to start in the `history` parameter
+-- @return true
+local function apply_history(rb, history, start)
+  
+  for i = start, #history do 
+    local target = history[i]
+    if target.weight > 0 then
+      assert(rb:addHost(target.name, target.port, target.weight))
+    else
+      assert(rb:removeHost(target.name, target.port))
+    end
+    rb.__targets_history[i] = {
+      name = target.name,
+      port = target.port,
+      weight = target.weight,
+      order = target.order,
+    }
   end
-  target.ip = ip
-  target.port = port
+  
   return true
 end
 
+-- looks up a balancer for the target.
+-- @param target the table with the target details
+-- @return balancer if found, or `false` if not found, or nil+error on error
+local get_balancer = function(target)
+  -- NOTE: only called upon first lookup, so `cache_only` limitations do not apply here
+  local hostname = target.upstream.host
+  
+  -- first go and find the upstream object, from cache or the db
+  local upstream, err = get_upstream(hostname)
+  if err then
+    return nil, err  -- there was an error
+  elseif upstream == false then
+    return false     -- no upstream by this name
+  end
+  
+  -- we've got the upstream, now fetch its targets, from cache or the db
+  local targets_history, err = cache.get_or_set(cache.targets_key(upstream.id), 
+                               nil, load_targets_into_memory, upstream.id)
+
+  if err then
+    return nil, err
+  end
+
+  local balancer = balancers[upstream.name] -- always exists, created upon fetching upstream
+  
+  -- check history state
+  local __size = #balancer.__targets_history
+  local size = #targets_history
+  if __size ~= size or 
```

Comment:
hmmm.. not really I explicitly did it this way because the `__` ones are now clear to be the ones on the `balancer` object and the non-`__` are the ones just fetched we're comparing against.

renaming them can be done of course, but it wouldn't make the code more readable imo.